src/visidentifier/pocowsclient.cpp

```diff
@@ -0,0 +1,261 @@
+    }
+
+    const auto response = requestParams->mResponse;
+
+    LOG_DBG() << "Got server response: requestId = " << requestId.c_str() << ", response = " << response.c_str();
+
+    return {response.cbegin(), response.cend()};
+}
+
+void PocoWSClient::SendMessage(const ByteArray& message)
+{
+    if (message.empty()) {
+        return;
+    }
+
+    std::lock_guard lock(mMutex);
+
+    if (!mIsConnected) {
+        throw WSException("Not connected", AOS_ERROR_WRAP(aos::ErrorEnum::eFailed));
+    }
+
+    try {
+        mWebSocket->setSendTimeout(Poco::Timespan(mWSConfig.mWebSocketTimeoutMiliseconds));
+
+        const int len = mWebSocket->sendFrame(&message.front(), message.size(), Poco::Net::WebSocket::FRAME_TEXT);
+
+        LOG_DBG() << "Sent " << len << "/" << message.size() << " bytes.";
+    } catch (const std::exception& e) {
+        mWSClientErrorEvent.Set(WSClientEvent::EventEnum::FAILED, e.what());
+
+        throw WSException(e.what(), AOS_ERROR_WRAP(aos::ErrorEnum::eFailed));
+    }
+}
+
+PocoWSClient::~PocoWSClient()
+{
+    Close();
+    StopReceiveFramesThread();
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+void PocoWSClient::HandleResponse(const std::string& frame)
+{
+    try {
+        const auto object = UtilsJson::ParseJson(frame).extract<Poco::JSON::Object::Ptr>();
+
+        if (object.isNull()) {
+            return;
+        }
+
+        if (const auto action = object->get(cActionTagName); action == "subscription") {
+            mHandleSubscription(frame);
+
+            return;
+        }
+
+        const auto requestId = object->get(cRequestIdTagName).convert<std::string>();
+        if (requestId.empty()) {
+            throw AosException("invalid requestId tag received");
+        }
+
+        if (!mPendingRequests.SetResponse(requestId, frame)) {
+            mHandleSubscription(frame);
+        }
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << AosException(e.what(), AOS_ERROR_WRAP(aos::ErrorEnum::eFailed)).what();
+    }
+}
+
+void PocoWSClient::ReceiveFrames()
+{
+    LOG_DBG() << "PocoWSClient::ReceiveFrames has been started.";
+
+    try {
+        int  flags;
+        int  n;
+        char buffer[1024];
+
+        do {
+            n = mWebSocket->receiveFrame(buffer, sizeof(buffer), flags);
+
+            LOG_DBG() << "PocoWSClient::ReceiveFrames recived " << n << " bytes. Flags = " << flags;
+
+            if ((flags & Poco::Net::WebSocket::FRAME_OP_BITMASK) == Poco::Net::WebSocket::FRAME_OP_CLOSE) {
+                mWSClientErrorEvent.Set(WSClientEvent::EventEnum::FAILED, "got Close frame from server");
+
+                return;
+            }
+
+            if (n > 0) {
+                const std::string message(buffer, n);
+
+                HandleResponse(message);
```

Comment:
you can use `string_view` instead of allocate string every time

Reply:
std::string is the best choice here with regards to Poco JSON parser and general safety.